scripts/runtests.py

```diff
@@ -0,0 +1,132 @@
+    lines = [line for line in json_text.split("\n") if not line.strip().startswith("//")]
+    # strip trailing commas from the file
+    valid_json = re.sub(r",(\s*?[\}\]])", r"\1", "\n".join(lines))
+    return valid_json
+
+
+def _get_strict_params(stub_path: str) -> list[str]:
+    with open(_STRICTER_CONFIG_FILE) as file:
+        data = json.loads(_parse_jsonc(file.read()))
+    if stub_path in data["exclude"]:
+        return []
+    return ["-p", _STRICTER_CONFIG_FILE]
+
+
+if __name__ == "__main__":
+    try:
+        path = sys.argv[1]
+    except IndexError:
+        print("Missing path argument in format <folder>/<stub>", file=sys.stderr)
+        sys.exit(1)
+    path_tokens = Path(path).parts
+    assert len(path_tokens) == 2, "Path argument should be in format <folder>/<stub>"
+    folder, stub = path_tokens
+    assert folder in {"stdlib", "stubs"}, "Only the 'stdlib' and 'stubs' folders are supported"
+    stubtest_result: subprocess.CompletedProcess[bytes] | None = None
+    pytype_result: subprocess.CompletedProcess[bytes] | None = None
+
+    # Run formatters first. Order matters.
+    subprocess.run([sys.executable, "-m", "pycln", path, "--all"])
+    subprocess.run([sys.executable, "-m", "isort", path])
+    subprocess.run([sys.executable, "-m", "black", path])
+
+    flake8_result = subprocess.run([sys.executable, "-m", "flake8", path])
+
+    check_consistent_result = subprocess.run([sys.executable, "tests/check_consistent.py"])
+    check_new_syntax_result = subprocess.run([sys.executable, "tests/check_new_syntax.py"])
+
+    pyright_result = subprocess.run(
+        [sys.executable, "tests/pyright_test.py", path] + _get_strict_params(path), stderr=subprocess.PIPE
+    )
+    print(pyright_result.stderr.decode())
+    if b"error running npx" in pyright_result.stderr:
+        pyright_returncode = 0
+        pyright_skipped = True
+    else:
+        pyright_returncode = pyright_result.returncode
+        pyright_skipped = False
+
+    mypy_result = subprocess.run([sys.executable, "tests/mypy_test.py", path])
+    # If mypy failed, stubtest will fail without any helpful error
+    if mypy_result.returncode == 0:
+        if folder == "stdlib":
+            stubtest_result = subprocess.run([sys.executable, "tests/stubtest_stdlib.py", stub])
+        else:
+            stubtest_result = subprocess.run([sys.executable, "tests/stubtest_third_party.py", stub])
+    else:
+        print("Skipping stubtest since mypy failed.")
+
+    if sys.platform == "win32":
+        print("Skipping pytype on Windows. You can run the test with WSL.")
+    else:
+        pytype_result = subprocess.run([sys.executable, "tests/pytype_test.py", path])
+
+    if folder == "stdlib":
+        regr_test_result = subprocess.run([sys.executable, "tests/regr_test.py", "stdlib"], stderr=subprocess.PIPE)
+    else:
+        regr_test_result = subprocess.run([sys.executable, "tests/regr_test.py", stub], stderr=subprocess.PIPE)
```

Comment:
Question: is the goal for this script to be a "convenience script", to run the most important checks? Or is it to be an "exhaustive script": to run _all_ the checks?

If it's a "convenience script", then we maybe don't need `mypy_test.py` to run on _all_ Python versions -- that takes forever, and usually isn't necessary. We could run `python tests/mypy_test.py -p <version-I'm-running>`.

If it's an "exhaustive script", then we should run `python tests/regr_test.py --all`, so it runs on _all_ platforms/python versions.

Reply:
I started this script with the idea that if it passes, the CI should pass. It's the convenience of not having to think about how each individual test expect their arguments, and not having to wait on the CI to tell you it failed many minutes later.

It's a valid question and I don't have a definite answer for it. It could be both, but then we have to decide which goal is a priority (and doesn't require an extra argument, wanting to keep this as dead simple to use as possible).

Reply:
Note also that by default `mypy_test.py` only runs with `--platform <platform-I'm- running`, so if we want to make this script truly exhaustive, we should iterate through all platforms on `mypy_test.py` as well :)

Personally, I'd go for making this a convenience script. It will still pick up on 95% of errors that will be flagged by CI. And emulating _all_ parts of our CI locally just isn't possible (unless you want to write a script to spin up a virtual machine to run `stubtest_stdlib.py` on linux, Windows and macOS, as well as whatever system you're running on locally ;).

Reply:
Food for thought: stubtest may be run in a different environment than the CI. Which means the results won't always 100% match up (but may allows the user to complete non-linux os-specific stubs that the CI may fail on, like I did with PyAutogui). That being said, this probably won't be true long-term.

As for mypy, I was gonna make a similar comment as you did. Should we still let it run on all python versions? It doesn't take *that* long. If we choose a specific version, either the current one being execute, or 3.7 since it has the most errors (mainly 3.10 union syntax in type alias)

Reply:
My preference would be to run `mypy_test.py`, `regr_test.py` and `pyright_test.py` all with the same Python version -- I think that's much easier for users to remember, increasing the simplicity of the script.

And good shout on using 3.7! We should add a comment to note why we're using 3.7 as opposed to any other Python version.

Reply:
Note added to the readme that the oldest version supported by typeshed will be used. The script also shows it to the user. And added a comment in the script explaining why (without mentioning any specific issue so that the comment doesn't get outdated)